features/steps/attach.py

```diff
@@ -0,0 +1,119 @@
+import json
+import logging
+import time
+
+from behave import when
+
+from features.steps.contract import change_contract_endpoint_to_staging
+from features.steps.shell import (
+    then_i_verify_that_running_cmd_with_spec_exits_with_codes,
+    when_i_run_command,
+)
+from uaclient.util import DatetimeAwareJSONDecoder
+
+ERROR_CODE = "1"
+
+
+@when("I initiate the magic attach flow")
+def when_i_initiate_magic_attach(context):
+    when_i_run_command(
+        context=context,
+        command="pro api u.pro.attach.magic.initiate.v1",
+        user_spec="as non-root",
+    )
+
+    magic_attach_resp = json.loads(
+        context.process.stdout.strip(), cls=DatetimeAwareJSONDecoder
+    )
+
+    context.magic_token = magic_attach_resp["data"]["attributes"]["token"]
+
+
+@when("I revoke the magic attach token")
+def when_i_revoke_the_magic_attach_token(context):
+    when_i_run_command(
+        context=context,
+        command="pro api u.pro.attach.magic.revoke.v1 --args magic_token={}".format(  # noqa
+            context.magic_token
+        ),
+        user_spec="as non-root",
+    )
+
+
+@when("I wait for the magic attach token to be activated")
+def when_i_wait_for_magic_attach_token(context):
+    when_i_run_command(
+        context=context,
+        command="pro api u.pro.attach.magic.wait.v1 --args magic_token={}".format(  # noqa
```

Comment:
Can we also isolate the magic-attach steps in a separate file ?

Reply:
Will do!